server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java

```diff
@@ -0,0 +1,472 @@
+                    return getDelegate().termState();
+                }
+            };
+        }
+    }
+
+    private abstract static class LazyFilterTermsEnum extends BaseTermsEnum {
+        abstract TermsEnum getDelegate() throws IOException;
+
+        @Override
+        public SeekStatus seekCeil(BytesRef text) throws IOException {
+            return getDelegate().seekCeil(text);
+        }
+
+        @Override
+        public void seekExact(long ord) throws IOException {
+            getDelegate().seekExact(ord);
+        }
+
+        @Override
+        public BytesRef term() throws IOException {
+            return getDelegate().term();
+        }
+
+        @Override
+        public long ord() throws IOException {
+            return getDelegate().ord();
+        }
+
+        @Override
+        public int docFreq() throws IOException {
+            return getDelegate().docFreq();
+        }
+
+        @Override
+        public long totalTermFreq() throws IOException {
+            return getDelegate().totalTermFreq();
+        }
+
+        @Override
+        public PostingsEnum postings(PostingsEnum reuse, int flags) throws IOException {
+            return getDelegate().postings(reuse, flags);
+        }
+
+        @Override
+        public ImpactsEnum impacts(int flags) throws IOException {
+            return getDelegate().impacts(flags);
+        }
+
+        @Override
+        public BytesRef next() throws IOException {
+            return getDelegate().next();
+        }
+
+        @Override
+        public AttributeSource attributes() {
+            try {
+                return getDelegate().attributes();
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    static int bloomFilterSize(int maxDocs) {
+        // 10% saturation (i.e., 10 bits for each term)
+        final long numBits = maxDocs * 10L;
+        if (numBits > Integer.MAX_VALUE) {
+            return Integer.MAX_VALUE;
+        } else {
+            return (int) numBits;
+        }
+    }
+
+    static int numBytesForBloomFilter(int bloomFilterSize) {
+        return Math.toIntExact((bloomFilterSize + 7L) / 8L);
+    }
+
+    static int hashTerm(BytesRef br) {
+        return hashTerm(br.bytes, 0x9747b28c, br.offset, br.length);
+    }
+
+    /**
+     * This is a very fast, non-cryptographic hash suitable for general hash-based lookup.
+     * <p>
+     * The C version of MurmurHash 2.0 found at that site was ported to Java by Andrzej Bialecki (ab at getopt org).
+     * <p>
+     * The code from getopt.org was adapted by Mark Harwood in the form here as one of a pluggable
+     * choice of hashing functions as the core function had to be adapted to work with BytesRefs with
+     * offsets and lengths rather than raw byte arrays.
+     */
+    private static int hashTerm(byte[] data, int seed, int offset, int len) {
```

Comment:
Maybe switch to `StringHelper.murmurhash3_x86_32` instead, it should be both faster and simpler? (Though I agree with forking it so that we wouldn't break bw compat if this method changed in Lucene)